csi/controllerserver.py

```diff
@@ -255,6 +258,118 @@ def ControllerGetCapabilities(self, request, context):
                     "rpc": {
                         "type": capability_type("LIST_VOLUMES")
                     }
+                },
+                {
+                    "rpc": {
+                        "type": capability_type("EXPAND_VOLUME")
+                    }
                 }
             ]
         )
+
+    def ControllerExpandVolume(self, request, context):
+        """
+        Controller plugin RPC call implementation of EXPAND_VOLUME
+        """
+
+        start_time = time.time()
+        expansion_requested_pvsize = request.capacity_range.required_bytes
+
+        # Get existing volume size
+        existing_volume = search_volume(request.volume_id)
+        existing_pvsize = existing_volume.size
+
+        # pvtype = PV_TYPE_SUBVOL
+
+        # if KADALU_VERSION in ["0.5.0", "0.4.0", "0.3.0"]:
+        #     for vol_capability in request.volume_capabilities:
+        #         # using getattr to avoid Pylint error
+        #         single_node_writer = getattr(csi_pb2.VolumeCapability.AccessMode,
+        #                                     "SINGLE_NODE_WRITER")
+
+        #         if vol_capability.access_mode.mode == single_node_writer:
+        #             pvtype = PV_TYPE_VIRTBLOCK
+
+        pvtype = PV_TYPE_SUBVOL
+
+        single_node_writer = getattr(csi_pb2.VolumeCapability.AccessMode,
+                                             "SINGLE_NODE_WRITER")
+
+        if request.volume_capability.AccessMode == single_node_writer:
+            pvtype = PV_TYPE_VIRTBLOCK)
+
+        logging.debug(logf(
+            "Found PV type",
+            pvtype=pvtype,
+            capability=request.volume_capability
+        ))
+
+        # Add everything from parameter as filter item
+        filters = {}
+        for pkey, pvalue in request.parameters.items():
+            filters[pkey] = pvalue
+
+        logging.debug(logf(
+            "Filters applied to choose storage",
+            **filters
+        ))
+
+        # UID is stored at the time of installation in configmap.
+        uid = None
+        with open(os.path.join(VOLINFO_DIR, "uid")) as uid_file:
+            uid = uid_file.read()
+
+        host_volumes = get_pv_hosting_volumes(filters)
+        logging.debug(logf(
+            "Got list of hosting Volumes",
+            volumes=",".join(v['name'] for v in host_volumes)
+        ))
+
+        hostvol = mount_and_select_hosting_volume(host_volumes, expansion_requested_pvsize)
+        if hostvol is None:
+            errmsg = "No Hosting Volumes available, add more storage"
+            logging.error(errmsg)
+            context.set_details(errmsg)
+            context.set_code(grpc.StatusCode.RESOURCE_EXHAUSTED)
+            return csi_pb2.CreateVolumeResponse()
+
+        mntdir = os.path.join(HOSTVOL_MOUNTDIR, hostvol)
+
+        if request.vol_capability.BlockVolume:
+            logging.info(logf(
+                "Volume is of type 'block'"
+            ))
+            expand_volume(mntdir)
```

Comment:
Is this the way of handling BlockVolume?

Reply:
Almost right! but I believe we will handle both VIRTBLOCK and BlockVolume almost similarly! Hence, doing the `expand_volume()` after `update_virtblock_volume()` is better. (in the same code block).

Reply:
sure